Railsでテストを書くプラクティス課題の提出

unikounio opened the pull request:
標記の件について提出させていただきます。
お手数ですがご確認のほどよろしくお願いいたします。

unikounio committed:
環境構築をして、BookクラスとUserクラスのテストを書いた

unikounio committed:
Reportクラスのインスタンスメソッドのテストを書いた

unikounio committed:
Reportクラスのシステムテストを書いた

unikounio committed:
FactoryBotを導入

unikounio committed:
fixturesを削除

unikounio committed:
コメント機能と日報言及機能のテストを書いた

unikounio committed:
Rubocopの指摘による微修正

maedana commented on test/factories/comments.rb:
```diff
@@ -0,0 +1,15 @@
+
+FactoryBot.define do
+  factory :book_comment, class: Comment do
+    user { User.find_by(email: 'alice@example.com') }
```
何の前提もなく`FactoryBot.create(:book_comment)`したときにデータが作られるようにしたほうがいいですね。`'alice@example.com'`が存在することを前提するのはNGかなーと思います。こういうことをしたいのであれば、`FactoryBot.create(:book_comment, user: xxx)`と明示的に渡すようにすべきかなと。他にも複数ありそうですが、前提条件がなくても動作するFactoryになるようにしましょう

unikounio replied:
ありがとうございます！
ご指摘いただいた視点でfactoryファイルをすべて確認し、修正しました！

maedana commented on test/factories/books.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :cherry_book, class: Book do
```
この感じでfactoryを定義していくと、他のデータが増えるたびにfactoryが増えていきます。通常は

```ruby
factory :book do
  ...
end
```
としておいて詳細な値は利用時に引数で渡すものかなぁと思います。ただ、`trait`や`transient`はうまく活用出来るなら大いにしたらよいと思います。

他のfactoryについても同様です。

unikounio replied:
ありがとうございます！
今回はfactoryで準備しておきたい`book`データが１つだけだったので、`trait`や`transient`は設定せずにfactoryの名称を抽象化するのみとしました。
`trait`については`comments`と`reports`で活用してみました～

maedana commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,18 @@
+  test '#created_on' do
+    fixed_time = Time.zone.local(2023, 10, 12, 20, 0, 0)
+    @report.created_at = fixed_time
+    assert_equal fixed_time.to_date, @report.created_on
```
`fixed_time.to_date`と動的な値にせず、静的な値で評価したいですね。2023/10/12であることを確認したいわけですよね

unikounio replied:
ありがとうございます！
直接Date.newするようにしました！

maedana commented on test/test_helper.rb:
```diff
@@ -8,8 +8,15 @@ class ActiveSupport::TestCase
+    fill_in 'Eメール', with: 'alice@example.com'
+    fill_in 'パスワード', with: 'password'
+    click_on 'ログイン'
+    assert_text 'ログインしました。'
```
ここでは別にassertしなくてもいいんじゃないでしょうか

unikounio replied:
ありがとうございます！
このassertしない場合ログイン処理がうまくいかなくなるのですが、こちらに記載のあるTurboが原因かな〜と思ったのでassertを書いています。→[turbo\-railsでテストが落ちる？ \| FBC](https://bootcamp.fjord.jp/questions/1586)
別の理由でassertがない方が良いということでしょうか…？

maedana replied:
なるほどー、理解しました。OKです。

maedana commented on test/test_helper.rb:
```diff
@@ -8,8 +8,15 @@ class ActiveSupport::TestCase
   # Add more helper methods to be used by all tests here...
+  include FactoryBot::Syntax::Methods
+
+  def login_alice
```
他のユーザでログインしたいときに困るので汎用的にしましょう。どのユーザでログインするかは引数で渡したいですね。

unikounio replied:
ありがとうございます！
特定のユーザに限定しない形に修正し、使用時に引数でユーザ指定できるようにしました！

unikounio committed:
factoryファイルとloginメソッドの修正
- factoryをcreateする際に前提条件が不要になるようにした
- factoryの定義をtraitを使用して整理した
- 特定のユーザに依存していたloginメソッドを汎用化した

unikounio committed:
created_onメソッドのテストを静的な値と照合するよう修正

maedana commented on test/factories/users.rb:
```diff
@@ -1,13 +1,8 @@
-  factory :bob, class: User do
-    email { 'bob@example.com' }
+  factory :user do
+    email { 'foo@example.com' }
```
これだと`FactoryBot.create(:user)`を複数回呼ぶとユニークエラーに引っかかりそうです。

連番となるようにしておくと防げるので対応するとよいですよー。
https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md#sequences

unikounio replied:
ありがとうございます！
シーケンス機能を把握してませんでした😅便利ですね～✨

maedana commented on test/test_helper.rb:
```diff
@@ -11,11 +11,13 @@ class ActiveSupport::TestCase
-    create(:alice) unless User.exists?(email: 'alice@example.com')
+  # TODO: ログインユーザを限定せず、汎用的なメソッドにする
+  # ログインするユーザは引数で渡す
+  def login_as_user(email = 'foo@example.com', password = 'password')
```
イメージ的には引数はUserのインスタンスかなと思ってました。存在しないときにこのメソッド内で作るのはちょっとややこしいのかなーと。メアドやパスの引数間違えてても気づかずに新しいユーザ作られちゃいますしね。

パスワードは渡せるようにしたほうがよさそうではありますが、実質的に`passwod`固定だと思うので必要性が出ない限りは引数なしで完全固定値でもよいかなぁと思います。

unikounio replied:
ありがとうございます！
なるほど。最初にメソッド内で作成するよう書いたら、その方向で頭が固まってしまっていたよう思います💧
自然なメソッドに仕上がってきて気持ちいいです🌿✨

maedana commented on test/system/reports_test.rb:
```diff
@@ -4,8 +4,8 @@
-    @report = create(:report_alice)
-    login_alice
+    @report = create(:report)
+    login_as_user
```
@report.userのメアド、login_as_userのデフォルトメアドが同じであることを知っている前提のコードなので読みくく感じました。

たとえば
```ruby
user = create(:user)
@report = create(:report, user:)
login_as_user(user)
```
みたいになってれば作ったユーザでログインしてるんだな、ということが明確になると思うんですよね。今は暗黙的過ぎるかなと。いかがでしょうか。

BooksTestも同様の話がありそうですね

unikounio replied:
ありがとうございます！
`login_as_user`の引数をUserのインスタンスにすると、テスト本体も読みやすく書けるんですね😄

maedana commented on test/models/report_test.rb:
```diff
@@ -4,17 +4,17 @@
   test '#editable?' do
-    assert @report.editable?(User.find_by(email: 'alice@example.com'))
-    assert_not @report.editable?(create(:bob))
+    assert @report.editable?(User.find_by(email: 'foo@example.com'))
```
ここだけじゃないんですけど、`@report = create(:report)`の結果作られるユーザのメアドが`'foo@example.com'`なんだということはFactoryのコードまでみないとわからないので読みにくさを感じます。
`user`も別途明示的に作った上でそのユーザを指定する形で`report`を作るようにすると、ここでも`User.find_by`なんてしなくても作ったuserを引数にすればよいだけかなと思います。いかがでしょうか。

unikounio replied:
ありがとうございます！
対応させていただきました。
他のご指摘も踏まえて、全体的にUserインスタンスを活用するという意識が足りていなかったのだと感じました💧

maedana commented on test/models/user_test.rb:
```diff
@@ -4,10 +4,10 @@
   test '#name_or_email' do
-    user = create(:alice)
-    assert_equal 'alice@example.com', user.name_or_email
+    user = create(:user)
```
factoryのデフォルト実装が変わると失敗するテストになってますよね。factoryの実装が変わってもそのまま動くテストであってほしいかなと思います。

unikounio replied:
ありがとうございます！
仰る通り、Userの実装が変わったことでテストが落ちていました😅
メアドを`user`インスタンスから呼び出す形にしてみました～
テスト下段部分についても、このファイルだけだと`user.email`がなんだかわからない以上「Foo」が入ってる可能性も考えられるので、`user.name`を使用する形にしました。

maedana commented on test/system/comments_test.rb:
```diff
@@ -4,9 +4,10 @@
-    login_alice
+    @book = create(:book)
+    @report = create(:report)
+    @email = 'foo@example.com'
```
前述してますが、このへんもメアドでなく、Userを明示的に作っておくほうが読みやすいテストになりそうだなと思います。

unikounio replied:
ありがとうございます！
対応しました～
userインスタンスを利用するにあたり、commentデータの作成もsetup内に持ってこようかと思ったのですが、不要なデータの作成が多くなりそうだったので、各テスト内での作成としました。

unikounio committed:
Userインスタンスを活用して堅牢性と可読性を高めた

unikounio committed:
Userクラスのname_or_emailメソッドの期待値をUserインスタンスから呼び出すよう変更

maedana commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,11 @@
+    assert_equal user.email, user.name_or_email
+
+    user.name = 'Foo'
+    assert_equal user.name, user.name_or_email
```
検証値を動的にしてますけど、検証値は静的にしたほうがよいと思うんです。で、userを作る部分で明示的にemailやnameを指定して作るとよいかなと思います。同じuserを使いまわすこともやめて、`メアドあり、名前なし`のuserと、`メアドあり、名前あり`のuserをそれぞれ別々に作って静的値で検証するとわかりやすいかな、と思いました。

unikounio replied:
ありがとうございます！
確かに検証値が動的だと、目標物が動き回っているようでわかりにくいですし、気持ち悪さがありました😅
対応させていただきました。

maedana commented on test/test_helper.rb:
```diff
@@ -8,8 +8,14 @@ class ActiveSupport::TestCase
+  def login_as_user(user)
+    visit new_user_session_path
+    fill_in 'Eメール', with: user.email
+    fill_in 'パスワード', with: user.password
```
パスワードを動的値にしてますけど、`password`ってDB保存値じゃないはずです。引数のuserがたまたまpasswordを設定された状態のインスタンスになってるだけかなと。それこそ対象userをfindしなおしたらpasswordはnilになってるんじゃないですかね。なので`password`固定値にしておけばいいかも、って前回書いたのでした。

unikounio replied:
ありがとうございます！
本番環境におけるパスワードの扱いに考えが及んでいなかったようです。
意図を組めているのか若干不安ですが修正してみました😅

maedana replied:
修正内容はOKですが、念の為書いておきます。

> 本番環境におけるパスワードの扱い

devise利用なので`encrypted_password`というのが永続化されたデータで暗号化されているかなと。`pasword`と`password_confirmation`はそれぞれ永続化されないですね。
```
irb(main):002:0> User.last.password
  User Load (1.4ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> nil
```
となりますから。修正前の状態で動いていたのはFactoryBotで作ったユーザがそのまま引数で渡ってきているのでメモリ上に一時的に存在する
```ruby
irb(main):003:0> User.new(password: 'password').password
=> "password"
```
の値が使われてから、ということになるかなと思います。

unikounio replied:
丁寧なご説明ありがとうございます！
修正前は一次保存されていたプレーンテキストの`password`が残っていたからログインできていただけであって、本番環境での挙動としては考えにくいものになっていたんですね💧
DBに保存されている`encrypted_password`は暗号化されているので、基本的にこれを呼び出してテストで使うことはできないということでしょうか。
理解が深まりました😄

maedana replied:
> DBに保存されているencrypted_passwordは暗号化されているので、基本的にこれを呼び出してテストで使うことはできないということでしょうか。

そうです。正確には暗号化ではなくHash化というもので不可逆な変換がされます。したがってログインフォームに入力する値としては使えないですね。なお、普通の暗号化で復号できてしまうと暗号化されたDBと復号に必要なキーがセットで漏洩とかするとパスワード漏れるので危ないですねー。なのでパスワードのような値は通常はHash化して保存しておき、照合時はフォームの入力値を再Hash化してDB上のHash化済みの値と照合して同じならOKみたいな実装になってるかなと思います。もし興味があればdeviseの実装とか追いかけてみるのも面白いかもですね。

unikounio replied:
ありがとうございます！
Hashって聞くと`{'Tom' => 'cat' }`を想像してしまうのですが、Hash化という不可逆な変換の仕組みがあるんですね。
ご説明いただいたおかげでイメージがかなり具体的になりました！

unikounio committed:
検証値が動的だった箇所とDBからパスワードを取得していた箇所の修正
